src/render/src/macros.rs

```diff
@@ -45,59 +45,89 @@ macro_rules! gfx_buffer_struct {
-            $( $field: usize, )*
-            layout: $crate::handle::raw::DescriptorSetLayout<B>,
-            set: $crate::pso::RawDescriptorSet<B>,
-        }
+        pub mod $name {
+            #[allow(unused_imports)]
+            use super::*;
+            use $crate::{pso, handle, buffer, image};
+            use $crate::core;
 
-        impl<B: $crate::Backend> $crate::pso::Descriptors<B> for $name<B> {
-            fn from_raw(
-                layout: $crate::handle::raw::DescriptorSetLayout<B>,
-                set: $crate::pso::RawDescriptorSet<B>
-            ) -> Self {
-                let mut binding = 0;
-                let mut next_binding = || {let b = binding; binding += 1; b };
-                $name {
-                    $( $field: next_binding(), )*
-                    layout,
-                    set
-                }
+            pub struct Set<B: $crate::Backend> {
+                $( $field: usize, )*
+                layout: handle::raw::DescriptorSetLayout<B>,
+                raw: pso::RawDescriptorSet<B>,
             }
 
-            fn layout_bindings() -> Vec<$crate::core::pso::DescriptorSetLayoutBinding> {
-                let mut bindings = Vec::new();
-                $({
-                    let binding = bindings.len();
-                    bindings.push($crate::core::pso::DescriptorSetLayoutBinding {
-                        binding,
-                        ty: <$bind as $crate::pso::Bind>::desc_type(),
-                        count: <$bind as $crate::pso::Bind>::desc_count(),
-                        // TODO: specify stage
-                        stage_flags: $crate::core::pso::ShaderStageFlags::all(),
-                    });
-                })*
-                bindings
+            pub struct Data<B: $crate::Backend> {
+                $( $field: [Option<<$bind as pso::Bind<B>>::Handle>; <$bind as pso::BindDesc>::COUNT], )*
             }
-            
-            fn layout(&self) -> &B::DescriptorSetLayout { self.layout.resource() }
-            fn set(&self) -> &B::DescriptorSet { self.set.resource() }
-        }
 
-        impl<B: $crate::Backend> $name<B> {
-            $(
-                fn $field(&self) -> $crate::pso::DescriptorSetBindRef<B, $bind> {
-                    $crate::pso::DescriptorSetBindRef {
-                        set: self.set(),
-                        binding: self.$field,
-                        phantom: ::std::marker::PhantomData,
-                    }
+            impl<B: $crate::Backend> pso::Descriptors<B> for Set<B> {
+                type Data = Data<B>;
+
+                fn from_raw(
+                    layout: handle::raw::DescriptorSetLayout<B>,
+                    raw: pso::RawDescriptorSet<B>
+                ) -> (Self, Self::Data) {
+                    let mut binding = 0;
+                    let mut next_binding = || {let b = binding; binding += 1; b };
+                    (Set {
+                        $( $field: next_binding(), )*
```

Comment:
nit: heh, not sure what we are getting by moving `next_binding()` up as opposed to just inlining it here

Reply:
we avoid getting a warning for the last unused `+=` operation.